### PR TITLE
New: Add "user actions"

### LINKED ISSUE #2249: Add support for user actions in chromium puppeteer

### LINKED ISSUE #3019: [Bug] Get `connector-puppeteer` branch coverage back to 80
It's currently 75

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

Allow users to execute actions before or after navigating to the target
URL.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Fix #2249
Fix #3019

### CODE REVIEW

File: packages/connector-puppeteer/tests/actions.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
+// import { URL } from 'url';
+import * as path from 'path';
+
+import test from 'ava';
+// import * as sinon from 'sinon';
```
Comment (sarvaje) [inline packages/connector-puppeteer/tests/actions.ts]: Remove all commented code.

File: packages/connector-puppeteer/tests/actions.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        });
+
+        connector.close();
+    } catch (e) {
+        t.is(e.message, `User action "${actionPath}" doesn't export a member "action".`);
+    }
+});
+
+
+test(`[${name}] User action gets loaded if it does not have the right shape`, (t) => {
```
Comment (sarvaje) [inline packages/connector-puppeteer/tests/actions.ts]: I think the description is wrong, it should be `[${name}] User action gets loaded if it has the right shape`

File: packages/connector-puppeteer/tests/actions.ts (hunk 1 above)
Comment (molant) [reply packages/connector-puppeteer/tests/actions.ts]: Done

File: packages/connector-puppeteer/tests/actions.ts (hunk 2 above)
Comment (molant) [reply packages/connector-puppeteer/tests/actions.ts]: I've updated the titles of the tests. Hopefully they make more sense now and are correct :)

**Comment by molant** [comment]
@antross, @sarvaje can you please take a look at this?

File: packages/connector-puppeteer/README.md [hunk 3]
```diff
@@ -148,6 +148,100 @@ E.g.:
...
+module.exports = {
+    action: async (page, options) => {
+        // your actions here
+    }
+};
+```
+
+The parameters the function receives are:
+
+* `page`: The [puppeteer `Page`][puppeteer page] with the tag used to navigate
```
Comment (molant) [inline packages/connector-puppeteer/README.md]:
```suggestion
* `page`: The [puppeteer `Page`][puppeteer page] with the tab used to navigate
```

File: packages/connector-puppeteer/README.md [hunk 4]
```diff
@@ -148,6 +148,100 @@ E.g.:
...
+
+The following is an example of an action that does Basic HTTP Authentication,
+please look at the source code of `connector-puppeteer` for other built-in
+actions:
+
+```ts
+export const basicHTTPAuth = async (page: puppeteer.Page, config: ConnectorOptions) => {
+    if (!(config && config.auth &&
+        typeof config.auth.user === 'string' &&
+        typeof config.auth.password === 'string')) {
```
Comment (antross) [inline packages/connector-puppeteer/README.md]: I like this example, but we should probably include a second example that demonstrates using `options.actionOptions`. Also, we'll want to explain that this example is reading first-party configuration options do to it's work - you can't add your own options directly to `config` itself.

File: packages/connector-puppeteer/package.json [hunk 5]
```diff
@@ -62,7 +62,6 @@
   "main": "./dist/src/connector.js",
   "name": "@hint/connector-puppeteer",
   "nyc": {
-    "branches": 75,
```
Comment (antross) [inline packages/connector-puppeteer/package.json]: 🎉

File: packages/connector-puppeteer/src/lib/actions.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+export const group = (actions: ActionConfig[] = []): UserActions => {
+    const beforeTargetNavigation: Action[] = [];
+    const afterTargetNavigation: Action[] = [];
+
+    for (const actionConfig of actions) {
+        let loadedAction: { action: Action};
+        const pathToUserAction = resolve(process.cwd(), actionConfig.file);
+
+        try {
+            loadedAction = require(pathToUserAction) as any;
```
Comment (antross) [inline packages/connector-puppeteer/src/lib/actions.ts]: I don't think `as any` should be necessary here as `require` already returns an `any` type an `loadedAction` is already explicitly typed.
```suggestion
            loadedAction = require(pathToUserAction);
```

File: packages/connector-puppeteer/src/lib/actions.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * Loads and groups user provided actions by the moment they need
+ * to be executed (`beforeTargetNavigation` or `afterTargetNavigation`).
+ * @param actions The user configuration with the actions
+ */
+export const group = (actions: ActionConfig[] = []): UserActions => {
+    const beforeTargetNavigation: Action[] = [];
+    const afterTargetNavigation: Action[] = [];
+
+    for (const actionConfig of actions) {
+        let loadedAction: { action: Action};
```
Comment (antross) [inline packages/connector-puppeteer/src/lib/actions.ts]:
```suggestion
        let loadedAction: { action: Action };
```

File: packages/connector-puppeteer/src/lib/actions.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        const pathToUserAction = resolve(process.cwd(), actionConfig.file);
+
+        try {
+            loadedAction = require(pathToUserAction) as any;
+        } catch (e) {
+            throw new Error(`Couldn't load user action in "${pathToUserAction}".${e.message}`);
```
Comment (antross) [inline packages/connector-puppeteer/src/lib/actions.ts]:
```suggestion
            throw new Error(`Couldn't load user action in "${pathToUserAction}". ${e.message}`);
```

File: packages/connector-puppeteer/src/lib/authenticators.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        selector: string;
+    };
+};
+
+/**
+ * `connector-puppeteer` action to authenticate on a website using Basic HTTP Auth.
+ * To be used on `beforeTargetNavigation`.
+ */
+export const basicHTTPAuth = async (page: puppeteer.Page, config: ConnectorOptions) => {
+    if (!(config && config.auth &&
```
Comment (antross) [inline packages/connector-puppeteer/src/lib/authenticators.ts]: This is a little hard to read given the wrapped negation. Perhaps breaking this into multiple `if (condition) { return }` blocks would be a bit easier to follow?

```ts
if (!config || !config.auth) {
    return;
}
if (typeof config.auth.user !== 'string' || typeof config.auth.password !== 'string') {
    return;
}
```

File: packages/connector-puppeteer/src/lib/authenticators.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    await page.authenticate({ password: config.auth.password, username: config.auth.user });
+};
+
+/**
+ * `connector-puppeteer` action to authenticate on a website using a form.
+ * To be used on `afterTargetNavigation`.
+ */
+export const formAuth = async (page: puppeteer.Page, config: ConnectorOptions) => {
+    if (!(config && config.auth &&
```
Comment (antross) [inline packages/connector-puppeteer/src/lib/authenticators.ts]: Same here about using multiple if-return blocks.

File: packages/connector-puppeteer/README.md (hunk 4 above)
Comment (molant) [reply packages/connector-puppeteer/README.md]: Done

File: packages/connector-puppeteer/src/lib/actions.ts (hunk 6 above)
Comment (molant) [reply packages/connector-puppeteer/src/lib/actions.ts]: Done

File: packages/connector-puppeteer/src/lib/authenticators.ts (hunk 9 above)
Comment (molant) [reply packages/connector-puppeteer/src/lib/authenticators.ts]: Done

File: packages/connector-puppeteer/src/lib/authenticators.ts (hunk 10 above)
Comment (molant) [reply packages/connector-puppeteer/src/lib/authenticators.ts]: Done

**Comment by molant** [comment]
@antross this is ready for another review.

File: packages/connector-puppeteer/README.md [hunk 11]
```diff
@@ -148,9 +148,146 @@ E.g.:
...
+
+        await page.authenticate({
+            password: config.auth.password,
+            username: config.auth.user
+        });
+    }
+}
+```
+
+**Note:** This user action users `options.auth` which is already
```
Comment (antross) [inline packages/connector-puppeteer/README.md]:
```suggestion
**Note:** This user action uses `options.auth` which is already
```

File: packages/connector-puppeteer/README.md [hunk 12]
```diff
@@ -148,9 +148,146 @@ E.g.:
...
+        if (typeof config.auth.user !== 'string' || typeof config.auth.password !== 'string') {
+            return;
+        }
+
+        await page.authenticate({
+            password: config.auth.password,
+            username: config.auth.user
+        });
+    }
+}
```
Comment (antross) [inline packages/connector-puppeteer/README.md]: Looks like this is causing linting to fail
```suggestion
};
```

File: packages/connector-puppeteer/README.md [hunk 13]
```diff
@@ -148,9 +148,146 @@ E.g.:
...
+```
+
+```js
+module.exports = {
+    action: async (page, config) => {
+        const selector = config.actionsOptions.elementId;
+
+        await page.click(selector);
+    }
+}
```
Comment (antross) [inline packages/connector-puppeteer/README.md]: And linting is complaining here too.
```suggestion
};
```

**Pull request merged**